### PR TITLE
JaCoCo Integration

### PR DESCRIPTION
**Pull request by cedrickcooke**
Closed and re-opened now that Code Climate was relaxed

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -2,6 +2,14 @@ version: 2
 
 jobs:
   assemble:
+    environment:
+      GRADLE_ARGS: >-
```
Comment (twyatt) [inline .circleci/config.yml]: Fancy.

File: build.gradle [hunk 2]
```diff
@@ -22,6 +22,7 @@ buildscript {
     dependencies {
         classpath 'com.android.tools.build:gradle:3.1.4'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:${versions.kotlin}"
+        classpath 'com.dicedmelon.gradle:jacoco-android:0.1.4'
```
Comment (twyatt) [inline build.gradle]: nit: can we use a `versions` variable, please?

**Comment by cedrickcooke** [comment]
> Looks great, but can you revert the moving of the files into the directory hierarchies?

Unfortunately, no I cannot. _Kotlin_ doesn't need folders in those files, but _Code Climate_ does. If you look at one of the CircleCI builds (such as [this one](https://circleci.com/gh/JuulLabs-OSS/able/252)) it "succeeds" but reports a bunch of errors that look like this:
```
ERRO[0000] failed to read file src/main/java/com/juul/able/experimental/messenger/GattCallback.kt
open src/main/java/com/juul/able/experimental/messenger/GattCallback.kt: no such file or directory 
WARN[0000] Couldn't find file for path "src/main/java/com/juul/able/experimental/messenger/GattCallback.kt" from build/reports/jacoco/jacocoTestDebugUnitTestReport/jacocoTestDebugUnitTestReport.xml coverage data. Ignore if the path doesn't correspond to an existent file in your repo. 
```
Their logic requires source files to be in the package-folders 😭.

File: .circleci/config.yml (hunk 1 above)
Comment (cedrickcooke) [reply .circleci/config.yml]: [YAML is simple](https://yaml-multiline.info/) 🙃

**Comment by efueger** [comment]
@cedrickcooke - Emily here with Code Climate Support. Thanks for the email about this.

I think this should do the trick:
- the test reporter has a [JACOCO_SOURCE_PATH](https://github.com/codeclimate/test-reporter/blob/master/man/cc-test-reporter-format-coverage.1.md#environment-variables) you can use to set the path to your source files
- related conversation here: https://github.com/codeclimate/test-reporter/issues/243#issuecomment-456673215

Let me know if that covers your use case. 👌 Happy to help further.

**Comment by twyatt** [comment]
Superseded by #42.

**Comment by jessecollier** [comment]
I was asked to look at why this couldn't work on code climate.

TL'DR: Codeclimate + jacoco depend on package name being where the files live. For example: com.juul.package => com/juul/package.kt.

Kotlin has slightly augmented this to not require the full package name when the whole package has a root prefix. For example the root is `com.juul` and therefore the file is just `package.kt` without the path prefix.

You can read more here:
https://kotlinlang.org/docs/reference/coding-conventions.html#source-code-organization

After looking into the cc-test-reporter, I found that it's making the assumption that the path name is the full package name (including the root). This obviously does not work with Kotlin best practices.
https://github.com/codeclimate/test-reporter/blob/master/formatters/jacoco/jacoco.go#L63

Therefore both jacoco and codeclimate need to support the root package name to omit that from the path structure.

Related: https://github.com/jacoco/jacoco/issues/939

**Comment by yonigoldberg** [comment]
What about Cobertura?

**Comment by twyatt** [comment]
> What about Cobertura?

Cobertura fails with:

```
* Where:
Build file '/Users/travis/Projects/able/build.gradle' line: 59

* What went wrong:
A problem occurred evaluating root project 'able'.
> Failed to apply plugin [id 'net.saliman.cobertura']
   > Could not create an instance of type net.saliman.gradle.plugin.cobertura.CoberturaExtension_Decorated.
      > Failed to apply plugin [class 'org.gradle.language.base.plugins.LifecycleBasePlugin']
         > Declaring custom 'clean' task when using the standard Gradle lifecycle plugins is not allowed.
```

It also [does not support the latest versions of Kotlin](https://github.com/stevesaliman/gradle-cobertura-plugin/issues/161).

**Pull request closed without merging**